tree: implement && and ~ for bbox operators

otan opened the pull request:
I have not implemented `@` (within) yet, as it causes a sql.y
shift-reduce issue which I'm yet to resolve. This should be enough to go
on for now. I've kept the same names for now, as I'm not sure what's
better at this point (`~` doesn't really make sense as `RegMatch` and
`GeoContains`, better name?!).

The experimental flags will be added in a later PR.

Release note (sql change): Implement the `&&` and `~` operators for
comparing bounding box objects and geometries.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/52913)
<!-- Reviewable:end -->

otan committed:
tree: implement && and ~ for bbox operators

I have not implemented `@` (within) yet, as it causes a sql.y
shift-reduce issue which I'm yet to resolve. This should be enough to go
on for now. I've kept the same names for now, as I'm not sure what's
better at this point (`~` doesn't really make sense as `RegMatch` and
`GeoContains`, better name?!).

The experimental flags will be added in a later PR.

Release note (sql change): Implement the `&&` and `~` operators for
comparing bounding box objects and geometries.

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 5 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52913#-:-MExsnA922mAXx5XeY2y:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

otan commented:
bors r+
